contributors/design-proposals/daemonset-create-first-rolling-update.md

```diff
@@ -0,0 +1,182 @@
+	// ensuring that at most 130% of the desired final number of DaemonSet  pods are running at all
+	// times during the update.
+	// +optional
+	MaxSurge *intstr.IntOrString `json:"maxSurge,omitempty" protobuf:"bytes,2,opt,name=maxSurge"`
+}
+```
+
+Then an additional hook for this strategy will be added to the part of the
+reconciliation loop that runs when pod updates are needed. That hook will:
+
+1. Determine how many nodes can have new pods scheduled, according to the
+   number of extra nodes currently running more than one pod and the `maxSurge`
+   parameter.
+2. Determine how many nodes can have their previous-generation pods retired
+   once the replacement pods are scheduled and running.
+3. Instruct the controller-manager to add and delete pods on nodes based on the
+   results of (1) and (2).
+
+This is mechanically almost identical to the current `RollingUpdate` strategy,
+with the key differences being the direct creation of new pods on nodes (the
+`RollingUpdate` strategy only deletes pods up to the `maxUnavailable` threshold
+and then waits for the main reconciler create new pods to replace them).
+
+One minor complication is that the controller-manager's reconciliation loop
+actively tries to enforce the one-pod-per-node invariant, and work is underway
+to fix cases in which the invariant is not held
+[kubernetes/kubernetes#50477](https://github.com/kubernetes/kubernetes/issues/50477).
+However, this invariant will need to be relaxed when the `SurgingRollingUpdate`
+strategy temporarily creates extra pods.
+
+### Alternatives considered
+
+The `maxSurge` parameter could have been added to the existing `RollingUpdate`
+strategy (as is the case for Deployments). However, this would break validation
+backwards compatibility, since the `maxUnavailable` parameter currently
+requires `maxUnavailable` to be > 0, but this is not required with
+`SurgingRollingUpdate`.
+
+There are plans to create a `v1beta2` iteration of the apps API (enabled by
+default in 1.9+), so it may be possible to add the `maxSurge` strategy to
+`RollingUpdate` and deprecate `SurgingRollingUpdate` then, or make the change
+now since the `v1beta2` API is incoming. However, to be conservative it was
+decided to implement this as a separate strategy.
+
+For more background see https://github.com/kubernetes/kubernetes/issues/48841.
+
+### Considerations / questions
+
+1. How are `hostPort`s handled?
+
+They are not handled as part of this proposal. We can either:
+
+   1. Attempt to determine a mechanism to handle this (unclear, perhaps best
+      left as future work, i.e. prior to GA)
+   2. Note in the documentation that this may cause port conflicts that prevent
+      new pods from scheduling successfully (and therefore updates from
+      completing successfully)
+   3. Add a validation check that rejects the combination of `hostPort` and the
+      `SurgingRollingUpdate` strategy.
+
+2. How will the scheduler handle hostPort collisions?
```

Comment:
The pod will not be scheduled because of `PodFitsHostPorts` predicate; and kubelet will also reuse this predicates to handle multi-scheduler case.